### ISSUE TITLE
exclusive_scan is faster in sequential than parallel

### ISSUE THREAD

**Issue by jefftrull** [STEllAR-GROUP/hpx#3733]
## Expected Behavior

Parallel should be slower than sequential for small data sizes, then faster after some crossover point.

## Actual Behavior

Parallel is consistently slower than sequential for data sizes from 10 to 10000000

## Steps to Reproduce the Problem

Compile and run the attached code with Google Benchmark.

## Specifications

gbench hardware report:
```
Run on (8 X 3800 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 6144K (x1)
```
  - HPX Version: 1.2.1
  - Platform (compiler, OS): gcc 8.2.0, Ubuntu 18.10

## Benchmark results:
```
-----------------------------------------------------------
Benchmark                    Time           CPU Iterations
-----------------------------------------------------------
Sequential/10                4 ns          4 ns  151589055
Sequential/64               27 ns         27 ns   26228755
Sequential/512             215 ns        215 ns    3314081
Sequential/4096           1719 ns       1719 ns     426286
Sequential/32768         13536 ns      13535 ns      50881
Sequential/262144       111051 ns     111045 ns       4988
Sequential/2097152     1108953 ns    1108894 ns        652
Sequential/10000000    5183736 ns    5183256 ns        131
Parallel/10              60798 ns    2172331 ns       3530
Parallel/64             149103 ns     172653 ns      10000
Parallel/512            151103 ns     312761 ns       1841
Parallel/4096           130873 ns 7155866560 ns          1
Parallel/32768          158808 ns     151704 ns       5030
Parallel/262144         227019 ns     161383 ns       3287
Parallel/2097152       1642302 ns 7508294482 ns          1
Parallel/10000000      7278212 ns    7715326 ns        182
```

**Comment by jefftrull** [comment]
Looks like I can't simply attach code to this issue, so here's the MWE as inline text:

```c++
#include <random>

#include <hpx/hpx_init.hpp>
#include <hpx/hpx.hpp>
#include <hpx/include/parallel_numeric.hpp>
#include <hpx/include/parallel_algorithm.hpp>

#include <benchmark/benchmark.h>

int main(int argc, char* argv[])
{
    // process and remove gbench arguments
    benchmark::Initialize(&argc, argv);

    // By default this should run on all available cores
    std::vector<std::string> const cfg = {
        "hpx.os_threads=all"
    };

    // Initialize and run HPX
    return hpx::init(argc, argv, cfg);
}

template<typename ExePolicy>
void exs_bench(ExePolicy & ex, char const * name)
{
    using namespace hpx::parallel;
    using namespace hpx::util;

    benchmark::RegisterBenchmark(name, [&](benchmark::State &state) {
      // create and fill random vector of desired size
      std::random_device rnd_device;
      std::mt19937 mersenne_engine{rnd_device()};
      std::uniform_int_distribution<int> dist{1, 20};

      auto sz = state.range(0);
      std::vector<int> data(sz);
      std::generate(data.begin(), data.end(),
                    [&]() { return dist(mersenne_engine); });
      std::vector<int> result(sz + 1);
      for (auto _ : state) {
        exclusive_scan(ex, data.begin(), data.end(), result.begin(), 0);
        benchmark::DoNotOptimize(result);
      }
    })->Range(10, 10000000);
}

int hpx_main(int argc, char **argv)
{
    // process commandline options

    using namespace hpx::parallel;
    using namespace hpx::util;

    exs_bench(execution::seq, "Sequential");
    exs_bench(execution::par, "Parallel");

    benchmark::RunSpecifiedBenchmarks();

    return hpx::finalize();
}

```

**Comment by hkaiser** [comment]
@jefftrull Thanks for doing this analysis. Unfortunately, we're not the only one experiencing problems in actually speeding up any of the scan algorithms an conventional cores. Would you have any suggestions on how to improve our implementation?

**Comment by jefftrull** [comment]
Ah... I see. No, it sounds pretty hard to me too :) I might do some profiling and get back to you, though. Thanks for the explanation.

**Comment by biddisco** [comment]
This seems very fishy - admittedly it was a couple of years ago, but when I implemented the vtk-m HPX backend, I tested the scan stuff quite extensively and I'm sure there was some speedup. I did comparisons with TBB and still have the graphs somewhere. I hope we haven't broken something fundamental. I will try to make time to look into this.

**Comment by biddisco** [comment]
Q1 : Does the benchmark include the setup time, or just the scan itself?
Q2 : what is the #include "benchmark.hpp" thing? just a timing utility or something more sophisticated?
Q3 : do try testing with `--hpx:threads=cores` rather than `--hpx:threads=all` - because hyperthreads

**Comment by jefftrull** [comment]
Thanks! I appreciate you looking into this. It may be that I'm misusing the system somehow.

1. The benchmark tests the call to `exclusive_scan` alone. It's everything inside the `for (auto _ : state)` loop.
2. This is part of [the popular microbenchmarking framework from Google](https://github.com/google/benchmark). @sithhell said he was familiar with it; if it's trouble to use I can pack up some more detailed instructions.
3. I tried with `cores` and the results are similar

**Comment by jefftrull** [comment]
I've put my code on Github [here](https://github.com/jefftrull/hpx_scanalgs). `exsvp` is the target you want to run.

**Comment by hkaiser** [comment]
@thanks @jefftrull !

**Comment by jefftrull** [comment]
I hope we can keep this bug open. With any luck I can propose a fix... I've been looking at it, on a low priority.

**Comment by jefftrull** [comment]
After a lot of thought and measurement I have some interesting perf data to share. The performance degradation seems to be due to cache conflicts. For a vector of 32b ints, size 16777216, I have the following data:

| | IPC | LLC loads | LLC misses |
| - | -- | -- | -- |
| Sequential | 2.69 | 498,208 | 130,808 |
| 2 threads | 1.97 | 43,147,967 | 28,560,110 |
| 4 threads | 1.95 | 32,195,045 | 22,295,854 |

To me this suggests a problem with memory locality or working set size.

**Comment by jefftrull** [comment]
![fast_16777216](https://user-images.githubusercontent.com/762705/67447105-aa8a5400-f5c7-11e9-9a97-313499bc0678.png)

I hope this visualization will help explain what I'm observing. The Y axis represents the input data - 16777216 32b ints in chunks of 90000 (chosen by tuning). The X axis is the execution time of a single run of the benchmark. The red blocks are the first pass for a given chunk, while the orange blocks are the final pass.

You can see here that essentially all of the first passes are completed before any final pass work begins. As a result, the data for each chunk must be freshly reloaded into cache for the final pass. I suspect from this and the LLC load miss data above that the memory interface bandwidth is the performance limiter. Thus the sequential algorithm - which accesses each location only once - outperforms for large data sizes.

**Comment by hkaiser** [comment]
@jefftrull very insightful, thanks!

**Comment by jefftrull** [comment]
Thanks! At your request I did another experiment with f3 dataflow launched as async(thread_priority_high) instead of sync. The result was a modest improvement in runtime but a considerable difference in the scheduling, as you can see:
![fast_16777216_f3_async_highpri](https://user-images.githubusercontent.com/762705/67449798-37391000-f5d0-11e9-8576-23f82fc8184c.png)

**Comment by hkaiser** [comment]
@jefftrull thanks a lot! Now, some of the stage 3 tasks are run right away (as it should be), the majority of those however is run after a bunch of stage 1 tasks have run. I think what's happening is that the high priority setting is inherited to tasks that are spawned as dependencies. Would you be able to do the same again, but using `threads_priority_boost` instead of `threads_priority_high`. That will raise the priority of the scheduled task but will not inherit the higher priority to any dependent ones.

**Comment by jefftrull** [comment]
It's my understanding that the stage 1 tasks have no dependencies... can this explain why they get run in preference to the stage 3 async/high priority tasks?

**Comment by jefftrull** [comment]
![priority_boost](https://user-images.githubusercontent.com/762705/67516830-28953c00-f656-11e9-8ef4-96b6d69acf71.png)
Not seeing a big improvement with that unfortunately...

**Comment by jefftrull** [comment]
Per @hkaiser 's IRC request here's a sweep of chunk size vs. runtime for the above input size, with 4 cores:

| Chunk Size | Time (ns) |
| ---------- | --------- |
| 20000 | 12126435 |
| 30000 | 11679925 |
| 40000 | 11597158 |
| 50000 | 11565757 |
| 60000 | 11635396 |
| 70000 | 11393571 |
| 80000 | 11437018 |
| 90000 | 11289811 |
| 100000 |11358582 |
| 110000 |11268380 |
| 120000 |11344181 |
| 130000 |11270813 |
| 140000 |11376445 |
| 150000 |11496581 |
| 160000 |11331730 |
| 170000 |11387999 |
| 180000 |11507707 |
| 190000 |11488620 |
| 200000 |11473176 |
| (Sequential) | 10030114 |

**Comment by jefftrull** [comment]
For comparison, here's input size 524288, again with 4 cores. You can see it's much more responsive to chunk size:
| Chunk Size | Time (ns) |
| ---------- | --------- |
| 20000 |   265958 |
| 30000 |   251997 |
| 40000 |   228854 |
| 50000 |   217671 |
| 60000 |   215546 |
| 70000 |   205306 |
| 80000 |   209854 |
| 90000 |   217660 |
| 100000 |  233200 |
| 110000 |  236815 |
| 120000 |  225239 |
| 130000 |  219319 |
| 140000 |  245155 |
| 150000 |  230407 |
| 160000 |  242736 |
| 170000 |  241209 |
| 180000 |  239539 |
| 190000 |  244929 |
| 200000 |  241961 |
| (Sequential) | 294180 |

I'm providing this chunk size variation without following my usual steps of logging out and running on a (mostly) idle system so this is just to demonstrate the variation... If the details are important please let me know.

**Comment by jefftrull** [comment]
A further request was to show the priority of the threads when each chunk is run, i.e. with a color code.

I began work on this experiment, but found that:

1. `thread_priority_boost` for the stage 3 work results in _every_ chunk being run with priority 2 (normal)
2. `thread_priority_high` results in every stage 3 task being run with priority 5 (high) and every stage 1 with priority 2, _and yet stage 1 tasks are preferentially executed_, which is unexpected to say the least.

So at this time there is nothing to be gained by showing the priorities in some different color.

**Comment by hkaiser** [comment]
@jefftrull thanks Jeff. Both observations need more investigation. Apparently we have somehow managed to break the thread priority logic in our scheduler.

**Comment by msimberg** [comment]
2. sounds pretty bad, but for 1. note that with `thread_priority_boost` the thread will be added to the high priority queue, but the priority will be changed to normal. So it will in fact be treated as a high priority thread. AFAIK we have no way to check which *queue* a thread was in (except to some extent with our logging...). Even though you couldn't see the priorities change, was there a difference in execution (time or order) with `thread_priority_boost`? I would also try with much bigger chunk sizes (such that you end up with something on the order 1-10 chunks per worker thread).

**Comment by jefftrull** [comment]
Interesting observation re: boost priority - if there's a way to check the queue do LMK. I didn't see much difference in performance with `boost` vs `high` as you can see from the above pictures (total runtime is in the title of the plots).

Generally speaking increasing the chunk size worsened performance from the point I chose - it seems to be a minimum for runtime, which is why I picked it. My theory here (which is in dispute) is that L3 cache misses play a dominant role in the performance.

**Comment by jefftrull** [comment]
Just ran the numbers (on a non-idle system) and I'll have to amend my last comment there a bit - performance does seem to improve a little past my chosen chunk size of 90K but it's nearly flat. Here's the data:

| Chunk Size | Runtime (ns) |
| ---------------- | ------------- |
| 20000 |     12247119 |
| 40000 |     11644954 |
| 80000 |     11333145 |
| 160000 |    11194785 |
| 320000 |    11090851 |
| 640000 |    11068486 |
| 1280000 |   11067613 |
| 2560000 |   10946563 |
| (Sequential) | 9694301 |

**Comment by jefftrull** [comment]
Also @hkaiser is it possible that the priorities are working correctly and there is a performance issue with dependencies? i.e., perhaps the high priority tasks are not considered "ready" for some reason.

We do see that a handful of them at the beginning are executed preferentially...

**Comment by jefftrull** [comment]
Also the gaps between the stage 3 tasks are interesting - almost as if a single thread decides to switch to stage 3 tasks, then another, then another, making three "lines" on the right side of those plots.

**Comment by hkaiser** [comment]
@jefftrull would you be willing to share the test code and the script that generates the plots with us?

**Comment by hkaiser** [comment]
Also, the table above shows that the best (parallel) execution time is reached for a chunksize of 640000 (i.e. 2560000 / 4, you were using 4 cores, so one chunk per core). OTOH, the graphs you shared seem to have much smaller chunks than that. What chunk size did you use to create the plots?

**Comment by jefftrull** [comment]
Yep, I think I used the chunksize that was best for a smaller data set (524288), which was 90000; I can use 640000 instead if that's helpful. Note it's only a little faster (roughly 3%).

I'll be happy to share the code; I had to hack HPX a bit to insert tracepoints so I'll have to think about how to cleanly do this in a fork...

**Comment by biddisco** [comment]
@jefftrull - if you have a test, please upload it here, I have tracing/logging capabilities in my branch and can plot out some task related info given a simple test.

**Comment by jefftrull** [comment]
Yep see "MWE" above!

**Comment by jefftrull** [comment]
@hkaiser requested I make available the visualization stuff I've been using to understand exclusive_scan execution. My HPX branch is [here](https://github.com/jefftrull/hpx/tree/feature/parallel_alg_tracepoints) and the repo with my test code is the same one but [a new branch](https://github.com/jefftrull/hpx_scanalgs/tree/feature/tracepoints) - `exsvp` is still the binary you want to run. [analyze_trace.py](https://github.com/jefftrull/hpx_scanalgs/blob/feature/tracepoints/analyze_trace.py) will make some nice pictures, including a runtime histogram and the "broken bar" graphs I've been pasting in this thread. Linux is required. I'm happy to answer any questions.

**Comment by hkaiser** [comment]
@jefftrull thanks a lot for preparing this!

**Comment by jefftrull** [comment]
I've added a visualization of when the `f2` combining task gets executed. A diamond marks the time that the result of the `f1` stage of a given partition is combined with the `f2` stage of the previous partition, conditionally enabling the next partition's `f2` and `f3` stages to begin.

**Comment by jefftrull** [comment]
Here's a sample image from that new analysis. You can see the code effectively becomes serial about 2/3 of the way through execution: 
![with_stage2](https://user-images.githubusercontent.com/762705/69503694-4a8e1280-0ed1-11ea-9387-03b467ab89f9.png)

Some observations:

- Nearly all the first stage tasks are run prior to any stage 2 or 3 tasks. This creates a couple of performance impacts:
  1. The output region for a given chunk is evicted from the cache between stages 1 and 3 
  2. The available parallelism is reduced, resulting in only a single thread being used toward the end
- Additionally, the 2nd and 3rd stages are interleaved in execution due to the order they are created in the code, but each 3rd stage depends on a previous 2nd stage. Doing more 2nd stages in advance would enable 3rd stage parallelism.

**Comment by jefftrull** [comment]
As a comparison point, this is a run with all the f2 tasks constructed _before_ the f3 tasks so they take priority. It constitutes a best case for scheduling, IMO.
![best_case](https://user-images.githubusercontent.com/762705/69503774-267f0100-0ed2-11ea-81a3-62f1b70e1eef.png)
Unfortunately variation in execution times of the stage 1 tasks meant that most runs did not do as well. Still I think it's something to aim for.

**Comment by david8dixon** [comment]
@jefftrull can you provide your configuration (i.e. what you passed to cmake command-line)

**Comment by jefftrull** [comment]
Well, it's been a while but nothing particularly special - this is g++ 8.30 (a.k.a. CMAKE_CXX_COMPILER) with BOOST_ROOT pointing to my local build of 1.71. Is there a particular variable that interests you?

**Comment by hkaiser** [comment]
#5575 implemented an improvement. Closing this for now. Feel free to reopen if you have more ideas on how to improve things.
